### PR TITLE
Add creation expiry time

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 1]
```diff
@@ -23,6 +23,60 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     using SignedIdentifier = Blobs::Models::SignedIdentifier;
     using FileQueryArrowField = Blobs::Models::BlobQueryArrowField;
     using FileQueryArrowFieldType = Blobs::Models::BlobQueryArrowFieldType;
+
+    /**
+     * @brief An access control object.
+     */
+    struct Acl final
```
Comment (LarryOsterman) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:30]: A minor nit: An "acl" is an "access control list" - in other words, it's a collection of objects. This appears to be an individual object.

The objects which make up an ACL are called "Access Control Entries" or ACE.

So I believe that pedantically, this should be an `Ace`, not an `Acl`.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 2]
```diff
@@ -378,6 +432,27 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+   * @brief Optional parameters for #Azure::Storage::Files::DataLake::FileClient::UploadFrom.
+   */
+  struct ScheduleFileDeletionOptions final
+  {
+    /**
+     * The expiry time from the specified origin. Only work if ExpiryOrigin is
+     * ScheduleFileExpiryOriginType::RelativeToCreation or
+     * ScheduleFileExpiryOriginType::RelativeToNow.
+     */
+    Azure::Nullable<std::chrono::milliseconds> TimeToExpire;
```
Comment (LarryOsterman) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:447]: What is the difference between `TimeToExpire` and `ExpiresOn`? 

And why is `TimeToExpire` expressed in milliseconds? That's a surprising unit of time (if it's a relative time, I would expect it to be something closer to std::chrono::duration).

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 3]
```diff
@@ -653,27 +728,6 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     } TransferOptions;
   };
 
-  using ScheduleFileExpiryOriginType = Blobs::Models::ScheduleBlobExpiryOriginType;
-
```
Comment (LarryOsterman) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:617]: Why was this moved in the file?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 4]
```diff
@@ -373,6 +432,29 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+  {
+    /**
+     * The expiry time from the specified origin. Only work if ExpiryOrigin is
+     * ScheduleFileExpiryOriginType::RelativeToCreation or
+     * ScheduleFileExpiryOriginType::RelativeToNow.
+     */
+    Azure::Nullable<std::chrono::milliseconds> TimeToExpire;
+
+    /**
+     * The expiry time in RFC1123 format. Only work if ExpiryOrigin is
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:450]: doc comment fix if this was hand-written:
```suggestion
     * The expiry time in RFC1123 format. Only works if ExpiryOrigin is
```

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 2 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:447]: There are four types in ScheduleFileExpiryOriginType : 
<html>
<body>


Expiry Option | Description
-- | --
RelativeToCreation | Sets the expiry time relative to the file creation time, x-ms-expiry-time must be specified as the number of milliseconds to elapse from creation time.
RelativeToNow | Sets the expiry relative to the current time, x-ms-expiry-time must be specified as the number of milliseconds to elapse from now.
Absolute | x-ms-expiry-time must be specified as an absolute time in RFC 1123 Format.
NeverExpire | Sets the file to never expire or removes the current expiry time, x-ms-expiry-time must not to be specified.


</body>
</html>

And then,
**TimeToExpire** is a relative time to Expire from now/creationTime in miliseconds, It will work under **RelativeToCreation** and **RelativeToNow**  types. Since these types are specified as the number of milliseconds, I think using milliseconds is esier for users to understand the unit of time supported by the REST API.
**ExpiresOn** is a absolute time(RFC 1123) to Expire. It will work under **Absolute** type.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 3 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:617]: Because **CreatePathOptions** uses **ScheduleFileDeletionOptions** , so I moved  **ScheduleFileDeletionOptions** before **CreatePathOptions**.  
And both **ScheduleFileExpiryOriginType** and **ScheduleFileDeletionOptions** are parameters of ScheduleDeletion, so I think it is better to make **ScheduleFileExpiryOriginType** and **ScheduleFileDeletionOptions** define in the same place，this will be more readable.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 1 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:30]: This is a valuable suggestion, I also think `Ace` may be better,but this was moved from somewhere else and would be a breaking change to change its name.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 5]
```diff
@@ -380,6 +439,29 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+    Azure::Nullable<std::chrono::milliseconds> TimeToExpire;
+
+    /**
+     * The expiry time in RFC1123 format. Only works if ExpiryOrigin is
+     * ScheduleFileExpiryOriginType::Absolute.
+     */
+    Azure::Nullable<DateTime> ExpiresOn;
+  };
+
+  using SchedulePathDeletionOptions = ScheduleFileDeletionOptions;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:463]: Does ScheduleDeletion also work for directory?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp [hunk 6]
```diff
@@ -421,6 +503,39 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * Specifies the duration of the lease, in seconds, or InfiniteLeaseDuration for a lease that
+     * never expires. A non-infinite lease can be between 15 and 60 seconds. A lease duration cannot
+     * be changed using renew or change. Does not apply to directories.
+     */
+    Azure::Nullable<std::chrono::seconds> LeaseDuration;
+
+    /**
+     * Optional parameters to schedule the file for deletion.
+     */
+    SchedulePathDeletionOptions ScheduleDeletionOptions;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:537]: Do we need to specify origin type?

File: sdk/storage/azure-storage-files-datalake/src/rest_client.cpp [hunk 7]
```diff
@@ -114,6 +114,18 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+          if (var0.count("creationTime") != 0)
+          {
+            vectorElement2.CreatedOn
+                = Core::_internal::Win32FileTimeConverter::Win32FileTimeToDateTime(
+                    std::stoll(var0["creationTime"].get<std::string>()));
+          }
+          if (var0.count("expiryTime") != 0)
+          {
+            vectorElement2.ExpiresOn
+                = Core::_internal::Win32FileTimeConverter::Win32FileTimeToDateTime(
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/rest_client.cpp:126]: Expiry time is also in Win32 FILETIME format? Can you confirm?

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_client_test.cpp [hunk 8]
```diff
@@ -245,6 +245,22 @@ namespace Azure { namespace Storage { namespace Test {
...
+    auto bufferStream = std::make_unique<Azure::Core::IO::MemoryBodyStream>(
+        Azure::Core::IO::MemoryBodyStream(buffer));
+    auto properties1 = m_fileClient->GetProperties();
+
+    // Append with flush option
+    Files::DataLake::AppendFileOptions options;
+    options.Flush = true;
+    m_fileClient->Append(*bufferStream, 0, options);
+    auto properties2 = m_fileClient->GetProperties();
+    EXPECT_NE(properties1.Value.ETag, properties2.Value.ETag);
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_client_test.cpp:261]: Need to check file size to ensure it's actually flushed.

File: sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_client_test.cpp [hunk 9]
```diff
@@ -245,6 +245,22 @@ namespace Azure { namespace Storage { namespace Test {
...
+  {
+    const int32_t bufferSize = 4 * 1024; // 4KB data size
+    std::vector<uint8_t> buffer(bufferSize, 'x');
+    auto bufferStream = std::make_unique<Azure::Core::IO::MemoryBodyStream>(
+        Azure::Core::IO::MemoryBodyStream(buffer));
+    auto properties1 = m_fileClient->GetProperties();
+
+    // Append with flush option
+    Files::DataLake::AppendFileOptions options;
+    options.Flush = true;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/test/ut/datalake_file_client_test.cpp:258]: Also need to test `false` case.

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 6 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:537]: Yes, because ScheduleDeletion is not supported for a directory, I have tested for this.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 5 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:463]: No, it's only for file.

File: sdk/storage/azure-storage-files-datalake/src/rest_client.cpp (hunk 7 above)
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/src/rest_client.cpp:126]: Yes, and I has already added an unit test for this scenario(in DataLakeFileSystemClientTest#ListPaths)

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 6 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:537]: How do we specify it's relative to creation time or current time, or absolute time?

Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:537]: This is aligned with .Net.
If TimeToExpire has value, is relative to current time, because this is create operation, I think current time is also creation time.
If Expires on has value, is absolute time.
The code is following:
```
    AZURE_ASSERT_MSG(
        !(options.ScheduleDeletionOptions.ExpiresOn.HasValue()
          && options.ScheduleDeletionOptions.TimeToExpire.HasValue()),
        "ExpiresOn and TimeToExpire are mutually exclusive.");
    if (options.ScheduleDeletionOptions.ExpiresOn.HasValue())
    {
      protocolLayerOptions.ExpiryOptions
          = Files::DataLake::ScheduleFileExpiryOriginType::Absolute.ToString();
      protocolLayerOptions.ExpiresOn = options.ScheduleDeletionOptions.ExpiresOn.Value().ToString(
          Azure::DateTime::DateFormat::Rfc1123);
    }
    else if (options.ScheduleDeletionOptions.TimeToExpire.HasValue())
    {
      protocolLayerOptions.ExpiryOptions
          = Files::DataLake::ScheduleFileExpiryOriginType::RelativeToNow.ToString();
      protocolLayerOptions.ExpiresOn
          = std::to_string(options.ScheduleDeletionOptions.TimeToExpire.Value().count());
    }
```

Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:537]: Maybe I can add some comments about this if this is confusing?

File: sdk/storage/azure-storage-files-datalake/src/rest_client.cpp (hunk 7 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/rest_client.cpp:126]: If it doesn't expire, this field would be `"0"`, which, after being converted to `Azure::DateTime`, would be `1601-01-01`.
It's not intuitive and is subject to changes in datetime implementation.

We need to do something about it.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp (hunk 6 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_options.hpp:541]: I think it's fine.

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp [hunk 10]
```diff
@@ -276,6 +276,14 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
           protocolLayerOptionsCopy,
           _internal::WithReplicaStatus(context));
 
+      for (auto& path : response.Value.Paths)
+      {
+        if (path.ExpiresOn.HasValue() && path.ExpiresOn.Value() == DateTime(1601))
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:281]: Don't use 1601 here

**Review by Jinming-Hu — APPROVED** [review]
